controllers/kvmcluster_controller.go

```diff
@@ -0,0 +1,159 @@
+	"github.com/pkg/errors"
+)
+
+const (
+	KVMClusterController = "kvmcluster"
+)
+
+// KVMClusterReconciler reconciles a KVMCluster object
+type KVMClusterReconciler struct {
+	client.Client
+	Recorder record.EventRecorder
+	Scheme   *runtime.Scheme
+}
+
+//+kubebuilder:rbac:groups=infrastructure.cluster.x-k8s.io.infrastructure.cluster.x-k8s.io,resources=kvmclusters,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=infrastructure.cluster.x-k8s.io.infrastructure.cluster.x-k8s.io,resources=kvmclusters/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=infrastructure.cluster.x-k8s.io.infrastructure.cluster.x-k8s.io,resources=kvmclusters/finalizers,verbs=update
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+// TODO(user): Modify the Reconcile function to compare the state specified by
+// the KVMCluster object against the actual cluster state, and then
+// perform operations to make the cluster state reflect the state specified by
+// the user.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.8.3/pkg/reconcile
+func (r *KVMClusterReconciler) Reconcile(ctx context.Context, req ctrl.Request) (_ ctrl.Result, reconcileError error) {
+	log := ctrl.LoggerFrom(ctx)
+
+	defer metrics.CaptureMetrics()
+
+	// Fetch the KVMCluster instance
+	kvmCluster := &v1alpha4.KVMCluster{}
+	err := r.Get(ctx, req.NamespacedName, kvmCluster)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+
+	defer func() {
+		if reconcileError != nil {
+			r.Recorder.Event(kvmCluster, "Warning", "failed-to-reconcile", reconcileError.Error())
+		}
+	}()
+
+	// Fetch the Cluster.
+	cluster, err := util.GetOwnerCluster(ctx, r.Client, kvmCluster.ObjectMeta)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	if cluster == nil {
+		log.Info("Cluster Controller has not yet set OwnerRef")
+		return reconcile.Result{}, nil
+	}
+
+	if annotations.IsPaused(cluster, kvmCluster) {
+		log.Info("KVMCluster or linked Cluster is marked as paused. Won't reconcile")
+		return reconcile.Result{}, nil
+	}
+
+	log = log.WithValues("cluster", cluster.Name)
+
+	// Create the scope.
+	clusterScope, err := scope.NewClusterScope(scope.ClusterScopeParams{
+		Cluster:        cluster,
+		KVMCluster:     kvmCluster,
+		ControllerName: KVMClusterController,
+	})
+	if err != nil {
+		return reconcile.Result{}, errors.Errorf("failed to create scope: %+v", err)
+	}
+
+	// Always close the scope when exiting this function so we can persist any KVMCluster changes.
+	defer func() {
+		if err := clusterScope.Close(); err != nil && reconcileError == nil {
+			reconcileError = err
+		}
+	}()
+
+	// Handle deleted clusters
+	if !kvmCluster.DeletionTimestamp.IsZero() {
```

Comment:
`reconcileDelete` and `reconcileNormal` is a pattern that I saw both in [aws](https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/1a29aec7bb94d44a38e2df6a398a681b2ec06d34/controllers/awscluster_controller.go#L139-L144) and [azure](https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/master/controllers/azurecluster_controller.go#L193-L198) operators.